### PR TITLE
Add testing and fix `weights_only` load for quantized types and nn.Parameters with python attrs

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Adds the following to allowed globals for the `weights_only` unpickler
- [x] `torch._utils._rebuild_qtensor` and qtensor related types
- [x] `torch._utils._rebuild_parameter_with_state` (used deserializing a parameter that has user-defined attributes like `Param.foo`)

The remaining rebuild functions that have not been allowlisted are

- [x] `torch._utils._rebuild_wrapper_subclass` (allowlisted in above PR)
- [ ] `torch._utils._rebuild_device_tensor_from_numpy` 
- [ ] `torch._utils._rebuild_xla_tensor` (legacy)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124330

### CODE REVIEW

File: torch/_weights_only_unpickler.py [hunk 1]
```diff
@@ -92,6 +92,12 @@ def _get_allowed_globals():
         torch.int16,
         torch.int32,
         torch.int64,
+        # quantized dtypes
+        torch.qint8,
+        torch.qint32,
+        torch.quint8,
+        torch.quint4x2,
+        torch.quint2x4,
```
Comment (albanD) [inline torch/_weights_only_unpickler.py:100]: We should have that list somewhere shared. Maybe we should re-use `torch.storage._dtype_to_storage_type_map` ?

File: torch/_weights_only_unpickler.py [hunk 2]
```diff
@@ -106,9 +112,20 @@ def _get_allowed_globals():
...
+        torch.per_channel_affine,
+        torch.per_channel_symmetric,
+        torch.per_channel_affine_float_qparams,
+    ]:
+        rc[str(qt)] = qt
     # Rebuild functions
     for f in [
         torch._utils._rebuild_parameter,
+        torch._utils._rebuild_parameter_with_state,
+        torch._utils._rebuild_qtensor,
```
Comment (albanD) [inline torch/_weights_only_unpickler.py:128]: I guess you inspected all 7 functions here?

File: torch/_weights_only_unpickler.py (hunk 1 above)
Comment (mikaylagawarecki) [reply torch/_weights_only_unpickler.py]: This list only contains dtypes serialized with `rebuild_tensor_v2` added `torch.storage._new_types` for dtypes serialized with `_rebuild_tensor_v3`

**Review by albanD — APPROVED** [review]
Thanks!

**Pull request closed without merging**